File: internal/probe/bind_linux.go
//go:build linux

package probe

import (
	"net"
	"syscall"
)

// bindToDeviceControl returns a net.Dialer.Control func that applies
// SO_BINDTODEVICE before connect(2), so a TCP/HTTP probe genuinely
// exercises the named interface rather than whatever route the kernel
// would otherwise pick (design.md §1, "uplink probing correctness").
func bindToDeviceControl(iface string) func(network, address string, c syscall.RawConn) error {
	return func(_, _ string, c syscall.RawConn) error {
		var sockErr error
		if err := c.Control(func(fd uintptr) {
			sockErr = syscall.SetsockoptString(int(fd), syscall.SOL_SOCKET, syscall.SO_BINDTODEVICE, iface)
		}); err != nil {
			return err
		}
		return sockErr
	}
}

// bindIPConnToDevice applies SO_BINDTODEVICE to an already-open raw IP
// socket (used by the ICMP prober, which can't go through Dialer.Control).
func bindIPConnToDevice(conn *net.IPConn, iface string) error {
	raw, err := conn.SyscallConn()
	if err != nil {
		return err
	}
	var sockErr error
	if err := raw.Control(func(fd uintptr) {
		sockErr = syscall.SetsockoptString(int(fd), syscall.SOL_SOCKET, syscall.SO_BINDTODEVICE, iface)
	}); err != nil {
		return err
	}
	return sockErr
}

File: src/probe/bind_linux.rs
//! `SO_BINDTODEVICE` -- binding a socket to a specific network interface
//! (a Linux-only socket option) so a probe genuinely exercises the named
//! NIC's route rather than whatever the kernel's default route selection
//! would otherwise pick. Matters specifically for uplink health-checking on
//! multi-homed hosts (wired/wireless/cellular): an unbound probe could
//! succeed via the wrong interface and mask that the "real" one is dead.
//!
//! This is the one piece the port spec calls out as highest-risk to get
//! subtly wrong. The Go original needs two different call shapes for this
//! (`net.Dialer.Control`'s pre-connect raw-fd callback for TCP/HTTP, vs
//! `*net.IPConn.SyscallConn()`'s post-open raw-fd callback for the ICMP raw
//! socket) because the two paths obtain a raw fd at different points in
//! their respective APIs. `socket2::Socket` gives direct access to the fd
//! throughout its lifetime, so both cases collapse to the same one
//! function here: bind the device before `connect()`/`send_to()` in both
//! callers (`tcp.rs` binds before `connect_timeout`; `icmp.rs` binds
//! before the first `send_to`), which is functionally identical to the Go
//! original's "before connect(2) happens" / "against an already-open
//! socket" split -- SO_BINDTODEVICE is idempotent-safe to apply any time
//! before the socket is used for I/O, on both a not-yet-connected TCP
//! socket and a connectionless raw socket.

use std::io;

use socket2::Socket;

/// Applies `SO_BINDTODEVICE` to `socket`, restricting it to `iface`.
/// Requires `CAP_NET_RAW` at runtime, exactly as the Go original --
/// granted via `modules/core.nix`'s `AmbientCapabilities` wiring whenever
/// `config.NeedsNetRaw()` is true, unaffected by this port.
pub fn bind_to_device(socket: &Socket, iface: &str) -> io::Result<()> {
    socket.bind_device(Some(iface.as_bytes()))
}
